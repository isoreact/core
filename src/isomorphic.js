import React from 'react';
import ReactDOMServer from 'react-dom/server';
import {first, map, shareReplay} from 'rxjs/operators';

import {ServerContext, BrowserContext} from './context';
import hasValueNow from './has-value-now';
import keyFor from './key-for';

/**
 * A function to create an RxJS <code>Observable</code>, optionally taking initial data when hydrating in the browser.
 *
 * @callback getData
 * @param {Object} props       - props passed to the isomorphic component, some of which it may use to look up data
 * @param {Object} [hydration] - initial data if we're in the browser and hydrating
 */

/**
 * Create an isomorphic version of a React component.
 *
 * @param {Object}        isomorphicComponent             - isomorphic component details
 * @param {String}        isomorphicComponent.name        - name
 * @param {Function}      isomorphicComponent.component   - React component
 * @param {React.Context} isomorphicComponent.context     - context to provide and consume the data stream
 * @param {getData}       isomorphicComponent.getData     - data stream creation function
 * @param {String}        isomorphicComponent.loadingProp - the name of the flag to toggle when loading/completed
 * @param {Object}        isomorphicComponent.propTypes   - propType validations
 * @returns {Function} the created isomorphic component
 */
export default function isomorphic({
    name,
    component:   Component,
    context:     Context,
    getData,
    loadingProp,
    propTypes, // eslint-disable-line react/forbid-foreign-prop-types
}) {
    const isoComponent = ({
        children, // eslint-disable-line no-unused-vars
        ...props
    }) => {
        const key = keyFor(name, props);

        if (process.browser) {
            return (
                <BrowserContext.Consumer>
                    {(getHydration) => (
                        <Context.Provider
                            value={{
                                data$: getData(props, getHydration(key), true).pipe(map(({props}) => props)),
                                loadingProp,
                            }}
                        >
                            <Component />
                        </Context.Provider>
                    )}
                </BrowserContext.Consumer>
            );
        } else {
            return (
                <ServerContext.Consumer>
                    {({getStream, registerStream}) => {
                        let stream$ = getStream(key);

                        if (!stream$) {
                            stream$ = getData(props, undefined, false);
                            registerStream(key, stream$);
                        }

                        // If the stream is resolved, render it.
                        if (hasValueNow(stream$)) {
                            return (
                                <Context.Provider
                                    value={{
                                        data$: stream$.pipe(map(({props}) => props), shareReplay(1)),
                                        loadingProp,
                                    }}
                                >
                                    <Component />
                                </Context.Provider>
                            );
                        }

                        // When the stream resolves later, continue walking the tree. (Also, the registerStream function can detect this condition.)
                        stream$
                            .pipe(first())
                            .toPromise()
                            .then(() => {
                                ReactDOMServer.renderToStaticMarkup(
                                    <ServerContext.Provider value={{getStream, registerStream}}>
                                        <Context.Provider
                                            value={{
                                                data$: stream$.pipe(map(({props}) => props), shareReplay(1)),
                                                loadingProp,
                                            }}
                                        >
                                            <Component />
                                        </Context.Provider>
                                    </ServerContext.Provider>
                                );
                            })
                            .catch(() => {
                                // This will also be caught in renderToHtml and is better handled there.
                            });

                        return null;
                    }}
                </ServerContext.Consumer>
            );
        }
    };

    isoComponent.__isomorphic_name__ = name;

    if (process.env.NODE_ENV === 'development') {
        isoComponent.propTypes = propTypes;
    }

    return isoComponent;
}
