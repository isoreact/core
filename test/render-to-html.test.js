import React from 'react';
import {__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS} from 'styled-components';

import renderToHtml from '../src/render-to-html';
import StyledComponentsServerRenderer from '../src/styled-components-server-renderer';

import {IsoSimple} from './data/isomorphic/iso-simple';
import {IsoNested} from './data/isomorphic/iso-nested';
import {IsoNestedWithStyles} from './data/isomorphic/iso-nested-with-styles';
import {IsoThrowsError} from './data/isomorphic/iso-throws-error';

import * as fetchBaseValue from './data/streams/fetch-base-value';
import * as fetchV from './data/streams/fetch-v';
import * as fetchW from './data/streams/fetch-w';

jest.mock('cuid');
require('cuid').mockImplementation(() => '0123456789abcdef');

const {StyleSheet} = __DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS;

describe('renderToHtml(isomorphicComponent)', () => {
    let originalProcessBrowser;

    beforeEach(() => {
        originalProcessBrowser = process.browser;
        process.browser = false;
    });

    afterEach(() => {
        process.browser = originalProcessBrowser;
    });

    describe('simple isomorphic component', () => {
        let fetchBaseValueSpy;
        let html;

        beforeEach(() => {
            fetchBaseValueSpy = jest.spyOn(fetchBaseValue, 'default');
        });

        afterEach(() => {
            html = undefined;
            fetchBaseValueSpy.mockRestore();
        });

        describe('with mount point className', () => {
            beforeEach(async () => {
                html = {body: await renderToHtml(<IsoSimple power={4} />, {className: 'mount-point'})};
            });

            test('renders correctly', () => {
                expect(html).toMatchSnapshot();
            });

            test('calls fetchBaseValue() once', () => {
                expect(fetchBaseValueSpy.mock.calls).toHaveLength(1);
            });
        });

        describe('without mount point className', () => {
            beforeEach(async () => {
                html = {body: await renderToHtml(<IsoSimple power={4} />)};
            });

            test('renders correctly', () => {
                expect(html).toMatchSnapshot();
            });

            test('calls fetchBaseValue() once', () => {
                expect(fetchBaseValueSpy.mock.calls).toHaveLength(1);
            });
        });
    });

    describe('nested isomorphic component', () => {
        let fetchBaseValueSpy;
        let fetchVSpy;
        let fetchWSpy;
        let html;

        beforeEach(async () => {
            fetchBaseValueSpy = jest.spyOn(fetchBaseValue, 'default');
            fetchVSpy = jest.spyOn(fetchV, 'default');
            fetchWSpy = jest.spyOn(fetchW, 'default');
            html = {body: await renderToHtml(<IsoNested coefficient={9} />)};
        });

        afterEach(() => {
            html = undefined;
            fetchWSpy.mockRestore();
            fetchVSpy.mockRestore();
            fetchBaseValueSpy.mockRestore();
        });

        test('renders correctly', () => {
            expect(html).toMatchSnapshot();
        });

        test('calls fetchBaseValue() once', () => {
            expect(fetchBaseValueSpy.mock.calls).toHaveLength(1);
        });

        test('calls fetchVSpy() once', () => {
            expect(fetchVSpy.mock.calls).toHaveLength(1);
        });

        test('calls fetchWSpy() once', () => {
            expect(fetchWSpy.mock.calls).toHaveLength(1);
        });
    });

    describe('styled nested isomorphic component', () => {
        beforeEach(() => {
            // Pretend we're on a server
            StyleSheet.reset(true);
        });

        afterEach(() => {
            // Stop pretending
            StyleSheet.reset();
        });

        test('renders correctly', async () => {
            const renderer = new StyledComponentsServerRenderer();
            const body = await renderToHtml(
                <IsoNestedWithStyles coefficient={9} />,
                {render: new StyledComponentsServerRenderer().render}
            );

            expect({head: renderer.getStyleTags(), body})
                .toMatchSnapshot();
        });
    });

    describe('non-isomorphic component', () => {
        describe('with mount point className', () => {
            test('renders as a regular element', async () => {
                expect({body: await renderToHtml(<div>Hello ;)</div>)}).toMatchSnapshot();
            });
        });

        describe('without mount point className', () => {
            test('renders as a regular element', async () => {
                expect({body: await renderToHtml(<div>Hello ;)</div>, {className: 'mount-point'})}).toMatchSnapshot();
            });
        });
    });

    describe('getData Observable produces an error', () => {
        let originalConsoleError;
        let consoleErrorSpy;
        let error;

        beforeEach(async () => {
            originalConsoleError = console.warn;
            console.error = () => {};
            consoleErrorSpy = jest.spyOn(console, 'error');

            return renderToHtml(<IsoThrowsError/>).then(({body}) => body).catch((e) => {
                error = e;
            });
        });

        afterEach(() => {
            consoleErrorSpy.mockRestore();
            console.error = originalConsoleError;
        });

        test('propagates the error to the caller', () => {
            expect(error).toBe('Goodnight');
        });
    });
});
