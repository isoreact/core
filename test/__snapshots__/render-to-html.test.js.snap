// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToHtml(isomorphicComponent) nested isomorphic component renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>
      <ul>
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__ISO_DATA__", "iso-nested", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {};
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "iso-nested--55fa1adb27dccf36b9460eaeb0421f3141422e12b248273991e4727dc2645cb0": {
          "v": 3,
          "w": 8
        },
        "iso-simple--cfa18ab5e5bfacba47b0cf02f2721b037b0e74dc6fcedc1f0fac019a83c36d6d": {
          "baseValue": 5
        }
      }
    });
  </script>
</body>
`;

exports[`renderToHtml(isomorphicComponent) non-isomorphic component with mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div>
    <div data-reactroot="">Hello ;)</div>
  </div>
</body>
`;

exports[`renderToHtml(isomorphicComponent) non-isomorphic component without mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div class="mount-point">
    <div data-reactroot="">Hello ;)</div>
  </div>
</body>
`;

exports[`renderToHtml(isomorphicComponent) simple isomorphic component with mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef" class="mount-point">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__ISO_DATA__", "iso-simple", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {};
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "iso-simple--cfa18ab5e5bfacba47b0cf02f2721b037b0e74dc6fcedc1f0fac019a83c36d6d": {
          "baseValue": 5
        }
      }
    });
  </script>
</body>
`;

exports[`renderToHtml(isomorphicComponent) simple isomorphic component without mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__ISO_DATA__", "iso-simple", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {};
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "iso-simple--cfa18ab5e5bfacba47b0cf02f2721b037b0e74dc6fcedc1f0fac019a83c36d6d": {
          "baseValue": 5
        }
      }
    });
  </script>
</body>
`;

exports[`renderToHtml(isomorphicComponent) styled nested isomorphic component renders correctly 1`] = `
<head>
  <style data-styled="lmVDjZ eNTiKw" data-styled-version="4.1.3">
    /* sc-component-id: StyledSection-semgqu-0 */
.lmVDjZ{padding:7px;background:#bbb;}
/* sc-component-id: StyledList-semgqu-1 */
.eNTiKw{margin:7px;background:#666;color:#ddd;}</style>
</head>

<body>
  <div id="0123456789abcdef">
    <section class="StyledSection-semgqu-0 lmVDjZ">
      <ul class="StyledList-semgqu-1 eNTiKw">
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__ISO_DATA__", "iso-nested-with-styles", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {};
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "iso-nested-with-styles--55fa1adb27dccf36b9460eaeb0421f3141422e12b248273991e4727dc2645cb0": {
          "v": 3,
          "w": 8
        },
        "iso-simple--cfa18ab5e5bfacba47b0cf02f2721b037b0e74dc6fcedc1f0fac019a83c36d6d": {
          "baseValue": 5
        }
      }
    });
  </script>
</body>
`;
